rust/hexane/src/cursor.rs

```diff
@@ -0,0 +1,590 @@
+            count,
+            value: Some(Cow::Owned(value)),
+        }
+    }
+}
+
+impl Run<'_, i64> {
+    pub fn delta(&self) -> i64 {
+        self.count as i64 * self.value.as_deref().cloned().unwrap_or(0)
+    }
+}
+
+impl<'a, T: Packable + ?Sized> Run<'a, T> {
+    pub fn new(count: usize, value: Option<Cow<'a, T>>) -> Self {
+        Run { count, value }
+    }
+
+    pub fn plus(mut self, num: usize) -> Self {
+        self.count += num;
+        self
+    }
+}
+
+pub trait ColumnCursor: Debug + Clone + Copy + PartialEq + Default {
+    type Item: Packable + ?Sized;
+    type State<'a>: EncoderState<'a, Self::Item>
+    where
+        <Self as ColumnCursor>::Item: 'a;
+    type PostState<'a>: Debug
+    where
+        Self::Item: 'a;
+    type Export: Debug + PartialEq + Clone;
+    type SlabIndex: Debug + Clone + HasPos + HasAcc + SpanWeight<Slab>;
+
+    // TODO: needs a test
+    fn encode<'a, I>(out: &mut Vec<u8>, values: I, force: bool) -> Range<usize>
+    where
+        I: Iterator<Item = Option<Cow<'a, Self::Item>>>,
+        Self::Item: 'a,
+    {
+        let start = out.len();
+        let mut state = Self::State::default();
+        for v in values {
+            state.append(out, v);
+        }
+        if !force && out.len() == start && state.is_empty() {
+            out.truncate(start);
+            return start..start;
+        }
+        state.flush(out);
+        let end = out.len();
+        start..end
+    }
+
+    fn empty() -> Self;
+
+    fn new(_: &Slab) -> Self {
+        Self::empty()
+    }
+
+    fn iter(slab: &[u8]) -> CursorIter<'_, Self> {
+        CursorIter {
+            slab,
+            cursor: Self::empty(),
+            run: None,
+        }
+    }
+
+    fn compute_min_max(_slabs: &mut [Slab]) {
+        for s in _slabs {
+            let (_run, c) = Self::seek(s.len(), s);
+            let _next = c.clone().next(s.as_slice());
+            assert!(_run.is_some());
+            assert!(_next.is_none());
+        }
+    }
+
+    fn is_empty(v: Option<Cow<'_, Self::Item>>) -> bool {
+        v.is_none()
+    }
+
+    fn contains(&self, run: &Run<'_, Self::Item>, agg: Agg) -> Option<Range<usize>> {
+        if agg == <Self::Item>::maybe_agg(&run.value) {
+            Some(0..run.count)
+        } else {
+            None
+        }
+    }
+
+    fn pop<'a>(&self, run: &mut Run<'a, Self::Item>) -> Option<Option<Cow<'a, Self::Item>>> {
+        run.next()
+    }
+
+    fn pop_n<'a>(
+        &self,
+        run: &mut Run<'a, Self::Item>,
+        n: usize,
+    ) -> Option<Option<Cow<'a, Self::Item>>> {
+        assert!(n > 0);
+        run.nth(n - 1)
+    }
+
+    // ENCODER
+    fn finalize_state<'a>(
+        slab: &'a Slab,
+        encoder: &mut Encoder<'a, Self>,
+        post: Self::PostState<'a>,
+        cursor: Self,
+    ) -> Option<Self>;
+
+    // ENCODER
+    fn finish<'a>(slab: &'a Slab, writer: &mut SlabWriter<'a, Self::Item>, cursor: Self);
+
+    fn copy_between<'a>(
```

Comment:
I think we should use a struct for the arguments of this function as I think it would be easier to read. I suggest:

```
struct CopyBetweenArgs<'a, C: ColumnCursor> {
    slab: &'a [u8],
    writer: &mut SlabWriter<'a, C::Item>,
    start: C,
    end: C,
    last_run: Run<'a, C::Item>,
}
```

This would make it easier to not confuse the start and end arguments and also make it clearer what `last_run` is for (assuming I've understood it correctly).